Implemented ordinary piece captures

8ru73u5 opened the pull request.

8ru73u5 committed:
Merge branch 'develop' of https://github.com/APRO-Project/app into feature-logic-piece-moves

8ru73u5 committed:
Add CaptureTree and Capture object to represent capture chains

Update cache type for allowedCapturesCache

8ru73u5 committed:
Merge branch 'develop' of https://github.com/APRO-Project/app into feature-piece-captures

8ru73u5 committed:
Implement calculating all possible captures that can be done by an ordinary piece

Deleted unnecessary CaptureTree and renamed Capture to CaptureChain

8ru73u5 committed:
Merge branch 'develop' of https://github.com/APRO-Project/app into feature-piece-captures

RouNNdeL committed:
Merge pull request #24 from APRO-Project/release-v1.0.3

Release v1.0.3

- Add simple piece movement
- Add hints for available moves
- Add crown icon to king pieces
- User interaction bug fixes

8ru73u5 committed:
Read from and write to allowed actions caches

Changed names of functions responsible for calculating allowed actions

8ru73u5 committed:
Include support for canCaptureBackwards option

Fix a bug when player could capture his own piece

8ru73u5 committed:
Merge branch 'master' of https://github.com/APRO-Project/app into feature-piece-captures

8ru73u5 committed:
Method to get capture final positions

Delete unnecessary checkIfEntryVisited method

8ru73u5 committed:
Make getCaptureFinalPositions method private and add getLongestCaptures method

8ru73u5 committed:
Allow move when capturing (don't remove pieces from grid yet)

RouNNdeL commented:
`getAllowedMoves` should also prioritize captures. There should not be any discrepancies between `getAllowedMoves` and `isMoveAllowed`.

RouNNdeL committed:
Remove finished TODO comment

RouNNdeL commented on app/src/main/java/com/cyberbot/checkers/game/Grid.java:
```diff
@@ -89,7 +87,15 @@ public GridEntry getEntryByCoords(int x, int y) throws IndexOutOfBoundsException
+        else return getAllowedMoves(src, true).contains(dst);
     }
 
     public boolean attemptMove(GridEntry src, GridEntry dst) {
```
The attempt move should no longer return whether or not the move was allowed, but whether or not the player can execute a next move.

RouNNdeL replied:
A good way to do it would probably be to internally split `moveAllowed` into `ordinaryMoveAllowed` and `captureAllowed`. The method would then expect a call to `moveAllowed` before `atemptMove` and if an attempt is made to move into an invalid location throw an `InvalidMoveException`.

8ru73u5 committed:
Split moveAllowed into moveAllowed and captureAllowed

8ru73u5 committed:
Add getCapturedPieces method

8ru73u5 commented:
> `getAllowedMoves` should also prioritize captures. There should not be any discrepancies between `getAllowedMoves` and `isMoveAllowed`.

As its name implies it returns allowed **moves** and doesn't care for captures. I agree though, that the name `moveAllowed` is misleading and I've already splitted that method into `moveAllowed` and `captureAllowed` as you requested.

8ru73u5 committed:
Add captureLength property and checks for CaptureChain object being a capture root and a capture endpoint

RouNNdeL commented:
Well, a capture is also a move, so when we show all available moves to the user we should also be showing the captures. Thus is believe that this method should also return the captures.

RouNNdeL commented:
Furthermore the current approach vastly complicates the implementation for the UI and the AI departments. Since both parties will be required to distinguish between moves and captures, which are a form of move.

8ru73u5 commented:
How about method [`getMovableEntries()`](https://github.com/APRO-Project/app/issues/22), couldn't it return movable entries? I really want to be consitent within my method naming. I know that a *move* can represent captures as well but it's higher level of abstraction. On the lower level I have to distinguish between those two actions and I believe UI and AI departments should have to separate them too, for example for the sake of removing captured pieces from the `Grid` or getting the longest capture. And what about introducing a new method called `actionAllowed()` that will do exactly what you want without interfering with current nomenclature?

8ru73u5 committed:
Add getCaptureRoot method

Remove check for being capture endpoint in checkIfEntryCaptured

Add getter for captureLength

8ru73u5 committed:
Add getMovableEntries method

RouNNdeL commented:
When merging this PR please make sure to include the string `Closes #22` in the commit message in order to automatically close the issue.

RouNNdeL commented:
When merging this PR please make sure to include the string `Closes #22` in the commit message in order to automatically close the issue.
